File: guides/micronaut-security-session-database-authentication/java/src/main/java/example/micronaut/DelegatingAuthenticationProvider.java
/*
 * Copyright 2017-2023 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package example.micronaut;

import io.micronaut.core.annotation.Nullable;
import io.micronaut.http.HttpRequest;
import io.micronaut.scheduling.TaskExecutors;
import io.micronaut.security.authentication.AuthenticationException;
import io.micronaut.security.authentication.AuthenticationFailed;
import io.micronaut.security.authentication.AuthenticationProvider;
import io.micronaut.security.authentication.AuthenticationRequest;
import io.micronaut.security.authentication.AuthenticationResponse;
import reactor.core.publisher.FluxSink;
import reactor.core.publisher.Flux;
import reactor.core.scheduler.Scheduler;
import reactor.core.scheduler.Schedulers;
import org.reactivestreams.Publisher;
import jakarta.inject.Named;
import jakarta.inject.Singleton;

import java.util.List;
import java.util.concurrent.ExecutorService;

import static io.micronaut.security.authentication.AuthenticationFailureReason.ACCOUNT_EXPIRED;
import static io.micronaut.security.authentication.AuthenticationFailureReason.ACCOUNT_LOCKED;
import static io.micronaut.security.authentication.AuthenticationFailureReason.CREDENTIALS_DO_NOT_MATCH;
import static io.micronaut.security.authentication.AuthenticationFailureReason.PASSWORD_EXPIRED;
import static io.micronaut.security.authentication.AuthenticationFailureReason.USER_DISABLED;
import static io.micronaut.security.authentication.AuthenticationFailureReason.USER_NOT_FOUND;

@Singleton // <1>
class DelegatingAuthenticationProvider implements AuthenticationProvider<HttpRequest<?>> {

    private final UserFetcher userFetcher;
    private final PasswordEncoder passwordEncoder;
    private final AuthoritiesFetcher authoritiesFetcher;
    private final Scheduler scheduler;

    DelegatingAuthenticationProvider(UserFetcher userFetcher,
                                     PasswordEncoder passwordEncoder,
                                     AuthoritiesFetcher authoritiesFetcher,
                                     @Named(TaskExecutors.BLOCKING) ExecutorService executorService) { // <2>
        this.userFetcher = userFetcher;
        this.passwordEncoder = passwordEncoder;
        this.authoritiesFetcher = authoritiesFetcher;
        this.scheduler = Schedulers.fromExecutorService(executorService);
    }

    @Override
    public Publisher<AuthenticationResponse> authenticate(@Nullable HttpRequest<?> httpRequest,
                                                          AuthenticationRequest<?, ?> authenticationRequest) {
        return Flux.<AuthenticationResponse>create(emitter -> {
            UserState user = fetchUserState(authenticationRequest);
            AuthenticationFailed authenticationFailed = validate(user, authenticationRequest);
            if (authenticationFailed != null) {
                emitter.error(new AuthenticationException(authenticationFailed));
            } else {
                emitter.next(createSuccessfulAuthenticationResponse(user));
                emitter.complete();
            }
        }, FluxSink.OverflowStrategy.ERROR).subscribeOn(scheduler); // <3>
    }

    private AuthenticationFailed validate(UserState user, AuthenticationRequest<?, ?> authenticationRequest) {
        AuthenticationFailed authenticationFailed = null;
        if (user == null) {
            authenticationFailed = new AuthenticationFailed(USER_NOT_FOUND);

        } else if (!user.isEnabled()) {
            authenticationFailed = new AuthenticationFailed(USER_DISABLED);

        } else if (user.isAccountExpired()) {
            authenticationFailed = new AuthenticationFailed(ACCOUNT_EXPIRED);

        } else if (user.isAccountLocked()) {
            authenticationFailed = new AuthenticationFailed(ACCOUNT_LOCKED);

        } else if (user.isPasswordExpired()) {
            authenticationFailed = new AuthenticationFailed(PASSWORD_EXPIRED);

        } else if (!passwordEncoder.matches(authenticationRequest.getSecret().toString(), user.getPassword())) {
            authenticationFailed = new AuthenticationFailed(CREDENTIALS_DO_NOT_MATCH);
        }

        return authenticationFailed;
    }

    private UserState fetchUserState(AuthenticationRequest<?, ?> authRequest) {
        final Object username = authRequest.getIdentity();
        return userFetcher.findByUsername(username.toString()).orElse(null);
    }

    private AuthenticationResponse createSuccessfulAuthenticationResponse(UserState user) {
        List<String> authorities = authoritiesFetcher.findAuthoritiesByUsername(user.getUsername());
        return AuthenticationResponse.success(user.getUsername(), authorities);
    }
}

File: guides/micronaut-security-session-database-authentication/python/src/example/micronaut/delegating_authentication_provider.py
from jakarta.inject import Singleton
from micronaut.http import HttpRequest
from micronaut.security.authentication import (
    AuthenticationFailureReason,
    AuthenticationRequest,
    AuthenticationResponse,
)
from micronaut.security.authentication.provider import HttpRequestAuthenticationProvider

from .authorities_fetcher import AuthoritiesFetcher
from .password_encoder import PasswordEncoder
from .user_fetcher import UserFetcher
from .user_state import UserState


@Singleton  # <1>
class DelegatingAuthenticationProvider(HttpRequestAuthenticationProvider):
    def __init__(
        self,
        user_fetcher: UserFetcher,
        password_encoder: PasswordEncoder,
        authorities_fetcher: AuthoritiesFetcher,
    ):
        self.user_fetcher = user_fetcher
        self.password_encoder = password_encoder
        self.authorities_fetcher = authorities_fetcher

    def authenticate(
        self,
        http_request: HttpRequest,
        authentication_request: AuthenticationRequest,
    ) -> AuthenticationResponse:
        user = self.fetch_user_state(authentication_request)  # <2>
        failure = self.validate(user, authentication_request)
        if failure is not None:
            return AuthenticationResponse.failure(failure)
        return self.create_successful_authentication_response(user)  # <3>

    def validate(self, user: UserState | None, authentication_request):
        if user is None:
            return AuthenticationFailureReason.USER_NOT_FOUND
        if not user.enabled:
            return AuthenticationFailureReason.USER_DISABLED
        if user.accountExpired:
            return AuthenticationFailureReason.ACCOUNT_EXPIRED
        if user.accountLocked:
            return AuthenticationFailureReason.ACCOUNT_LOCKED
        if user.passwordExpired:
            return AuthenticationFailureReason.PASSWORD_EXPIRED
        if not self.password_encoder.matches(
            str(authentication_request.getSecret()),
            user.password,
        ):
            return AuthenticationFailureReason.CREDENTIALS_DO_NOT_MATCH
        return None

    def fetch_user_state(self, authentication_request):
        username = str(authentication_request.getIdentity())
        return self.user_fetcher.findByUsername(username).orElse(None)

    def create_successful_authentication_response(
        self,
        user: UserState,
    ) -> AuthenticationResponse:
        authorities = self.authorities_fetcher.findAuthoritiesByUsername(user.username)
        return AuthenticationResponse.success(user.username, authorities)
